### PR TITLE
improve `tfplan` files local storing process

### LINKED ISSUE #261: Rules for storing tfplan files locally
## Describe the Feature

Provide feature to store locally historical terraform.tfplan files

## Steps to Reproduce

1. `Default` workspace, no backup folder defined in `.terrahub.yml`
```
component/
   ├─ terraform.tfplan
   ├─ terraform.tfstate
   └─ .backup
       ├─ tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  └─ terraform.123456789012.tfplan
       └─ tfstate
          ├─ terraform.123456789012.tfstate
          ├─ terraform.123456789012.tfstate
          ├─ terraform.123456789012.tfstate
          └─ terraform.123456789012.tfstate
```

2. Workspace `dev`, no backup folder defined in `.terrahub.yml`
```
component/
   ├─ terraform.tfplan
   └─ terraform.tfstate
   └─ terraform.tfstate.d
      └─ dev
         ├─ terraform.tfplan
         ├─ terraform.tfstate
         └─ .backup
             ├─ tfplan
             │  ├─ terraform.123456789012.tfplan
             │  ├─ terraform.123456789012.tfplan
             │  ├─ terraform.123456789012.tfplan
             │  └─ terraform.123456789012.tfplan
             └─ tfstate
                ├─ terraform.123456789012.tfstate
                ├─ terraform.123456789012.tfstate
                ├─ terraform.123456789012.tfstate
                └─ terraform.123456789012.tfstate
```

3. `Default` workspace, backup folder is defined in `.terrahub.yml`
- .terrahub.yml
```yaml
terraform:
  version: 0.11.7
  backup: 'backup_folder'
```
- structure
```
component/
   ├─ terraform.tfplan
   ├─ terraform.tfstate
   └─ backup_folder
       ├─ tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  └─ terraform.123456789012.tfplan
       └─ tfstate
          ├─ terraform.123456789012.tfstate
          ├─ terraform.123456789012.tfstate
          ├─ terraform.123456789012.tfstate
          └─ terraform.123456789012.tfstate
```

4. Workspace `dev`, backup folder is defined in `.terrahub.yml`
```
component/
   ├─ terraform.tfplan
   ├─ terraform.tfstate
   ├─ terraform.tfstate.d
   │  └─ dev
   │     ├─ terraform.tfplan
   │     └─ terraform.tfstate
   └─ backup_folder
       ├─ tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  ├─ terraform.123456789012.tfplan
       │  └─ terraform.123456789012.tfplan
       ├─ tfstate
       │   ├─ terraform.123456789012.tfstate
       │   ├─ terraform.123456789012.tfstate
       │   ├─ terraform.123456789012.tfstate
       │   └─ terraform.123456789012.tfstate
       └─ dev
          ├─ tfplan
          │  ├─ terraform.123456789012.tfplan
          │  ├─ terraform.123456789012.tfplan
          │  ├─ terraform.123456789012.tfplan
          │  └─ terraform.123456789012.tfplan
          └─ tfstate
             ├─ terraform.123456789012.tfstate
             ├─ terraform.123456789012.tfstate
             ├─ terraform.123456789012.tfstate
             └─ terraform.123456789012.tfstate
```

### PR DESCRIPTION
**Pull request by Leichtwind**
## Description

Fixes #261 Rules for storing tfplan files locally

## Type of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] This change requires a change to the documentation

## How Has This Been Tested?

- [x] `terrahub run --apply --destroy` for local state
- [x] `terrahub run --apply --destroy --env dev` for local state
- [x] `terrahub run --apply --destroy` for remote state
- [x] `terrahub run --apply --destroy --env dev` for remote state
- [x] `terrahub run --apply --destroy` for local state and configured backup folder
- [x] `terrahub run --apply --destroy --env dev` for local state and configured backup folder
- [x] `terrahub run --apply --destroy` for remote state and configured backup folder
- [x] `terrahub run --apply --destroy --env dev` for remote state and configured backup folder

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have checked that my changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/helpers/terraform.js [hunk 1]
```diff
@@ -291,15 +299,13 @@ class Terraform {
...
   apply() {
     const params = {};
-    const planPath = this._plan.getPath();
-    const statePath = this._state.getPath();
+    const planPath = this._metadata.getPlanPath();
+    const statePath = this._metadata.getStatePath();
 
-    if (!this._state.isRemote()) {
+    if (!this._metadata.isRemote()) {
       if (fs.existsSync(statePath)) {
```
Comment (ddimitrioglo) [inline src/helpers/terraform.js]: Why do we need this check?

File: src/helpers/terraform.js [hunk 2]
```diff
@@ -291,15 +299,13 @@ class Terraform {
...
         Object.assign(params, {
-          '-state': statePath,
-          '-backup': this._state.getBackupPath(),
-          '-state-out': statePath
+          '-backup': this._metadata.getStateBackupPath()
         });
       } else if (fs.existsSync(planPath)) {
         Object.assign(params, { '-state-out': statePath });
```
Comment (ddimitrioglo) [inline src/helpers/terraform.js]: terraform makes this natively

File: src/helpers/terraform.js [hunk 3]
```diff
@@ -332,12 +338,12 @@ class Terraform {
    */
   destroy() {
     const options = {};
-    const statePath = this._state.getPath();
+    const statePath = this._metadata.getStatePath();
 
-    if (!this._state.isRemote() && fs.existsSync(statePath)) {
+    if (!this._metadata.isRemote() && fs.existsSync(statePath)) {
       Object.assign(options, {
         '-state': statePath,
```
Comment (ddimitrioglo) [inline src/helpers/terraform.js]: states and plans places are managed by terraform

**Comment by KanoniPan** [comment]
Next tests passed:
- `terrahub run --apply --destroy` for local state
- `terrahub run --apply --destroy --env` dev for local state
- `terrahub run --apply --destroy` for remote state
- `terrahub run --apply --destroy --env` dev for remote state
- `terrahub run --apply --destroy` for local state and configured backup folder
- `terrahub run --apply --destroy --env dev` for local state and configured backup folder
- `terrahub run --apply --destroy` for remote state and configured backup folder
- `terrahub run --apply --destroy --env dev` for remote state and configured backup folder

Every command worked as expexted.

TerraHub Version: v0.0.21
OS: MacOS High Sierra v10.13.6

**Pull request merged**